cpp/src/transform/transform.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/cudf.h>
+#include <jit/type.h>
+#include <nvstrings/NVCategory.h>
+#include <utilities/cudf_utils.h>
+#include <bitmask/legacy/bitmask_ops.hpp>
+#include <bitmask/legacy/legacy_bitmask.hpp>
+#include <cudf/copying.hpp>
+#include <string/nvcategory_util.hpp>
+#include <utilities/error_utils.hpp>
+#include "jit/core/launcher.h"
+
+#include <utilities/column_utils.hpp>
+
+namespace cudf {
+namespace transformation {
+
+/**---------------------------------------------------------------------------*
+ * @brief Computes output valid mask for op between a column and a scalar
+ *
+ * @param out_null_coun[out] number of nulls in output
+ * @param valid_out preallocated output mask
+ * @param valid_col input mask of column
+ * @param num_values number of values in input mask valid_col
+ *---------------------------------------------------------------------------**/
+
+namespace jit {
+
+void unary_operation(gdf_column* out, const gdf_column* in,
+                     const std::string& ptx, const std::string& output_type) {
+  Launcher(ptx, output_type).setKernelInst("kernel", out, in).launch(out, in);
+}
+
+}  // namespace jit
+
+}  // namespace transformation
+
+gdf_column transform(const gdf_column& input,
+                     const std::string& ptx_unary_function,
+                     gdf_dtype output_type) {
+  
+  // First create a gdf_column and then call the above function
+  gdf_column output = allocate_size_dtype(input.size, output_type, input.valid != nullptr);
+  output.null_count = input.null_count;
+
+  // Check for 0 sized data
+  if ((output.size == 0) && (input.size == 0)){
+      return output;
+  }
+
+  // Check for null data pointer
+  CUDF_EXPECTS((input.data != nullptr), "Input column data pointers are null");
+
+  // Check for datatype
+  // Input data types can be different but they have to be one of the
+  // following four.
+  CUDF_EXPECTS( input.dtype == GDF_FLOAT32 ||
```

Comment:
Hmmm, why not other data types? I can see why some types, like strings, are difficult. But what about STRING_CATEGORY? If someone wants to write a UDF that transforms the category index of each row, why shouldn't they? It's just an integer index. Then just do a nvcategory_gather on the output afterwards...

Reply:
I just tried and I could easily extend to include `int16`. For `GDF_STRING_CATEGORY` the only concern is that for `int32` type in most of the cases `numba` promotes it to `int64`, performs the operations and return a `int64`. This is certainly fine: we can always convert this back to `int32` manually. However I feel this is very far away from the original string manipulation purpose, not to mention that we are only transforming the indices instead of the strings themselves.

Reply:
For `int8` there is a technical difficulty here: the inline PTX does not seem to support 1-byte data types, i.e. there is no constraint letter to replace `x_reg` here (`input_param` is of type `int8_t`),
> asm volatile ("  cvt.u8.u8 _rs1,  %0;": : "x_reg"(input_param));
Certainly we could use the trick
> asm volatile ("  cvt.u8.u8 _rs1,  %0;": : "h"(static_cast<short>(input_param)));
But this might not be the solution we want?

Reply:
I think these are the 7 most accessible types in `libcudf`: `int8`, `int16`, `int32`, `int64`, `float32`, `float64`, `GDF_STRING_CATEGORY`.

Reply:
Surely inline PTX must be able to handle chars...

Reply:
Hmm, I guess you are right: https://stackoverflow.com/a/14289597/749748

Reply:
I think just do numerical types for now. And you should be able to cast 8-bit types as in the SO post and your comment above.

Reply:
I found that `date32`, `date64` and `gdf_timestamp` are also integer types so they can be added(?) I already have the working function to synchronize the new indices with the `nvcategory` object so `GDF_STRING_CATEGORY` is also supported.